### ISSUE TITLE
select_optimal_assignment_mask(reduce_matrix()) is behaving not as expected

### ISSUE THREAD

**Issue by lazure-ocean** [mmgalushka/hungarian-loss#7]
Hello! Thank you for your work.

Before using your tensorflow implementation of hungarian algorithm I ran several tests and noticed that on some inputs it does not work as expected (I was comparing it to `scipy.optimize.linear_sum_assignment`).

For instance, the following code
```
from hungarian_loss.steps import (
    reduce_matrix,
    select_optimal_assignment_mask,
)
import tensorflow as tf
import numpy as np

cost = np.array(
    [[0., 9., 7., 5.],
     [4., 2., 0., 9.],
     [7., 5., 4., 2.],
     [0., 9., 7., 5.]]
)
cost = tf.cast(tf.constant(cost), tf.float32)
select_optimal_assignment_mask(reduce_matrix(cost))
```

returns
```
<tf.Tensor: shape=(4, 4), dtype=bool, numpy=
array([[ True, False, False, False],
       [False, False,  True, False],
       [False,  True, False, False],
       [False, False, False, False]])>
```

As you can see, it matches only 3 pairs instead of 4. The last row and last column do not have any `True` value.

Do you have any ideas, why this is happenning and how to fix this? Or am I using it incorrectly?

**Comment by nviwch** [comment]
This is what I wanted for my loss function. It seems no one wrote linear sum assignment in Tensorflow, and scipy functionslowed down my training, and the gradient is missing.

Please fix the issue and also add support to non-squared cost matrix

**Comment by nviwch** [comment]
After debugging into your example, I found that the scratch_matrix function is incorrect, it did not return the **MINIMUM** number of lines that scratch zeros.

Briefly checked reduce_matrix and shift_zero seems correct as the logic is obvious, scratch_matrix is not.

It is another optimization problem finding the minimum number of lines. In the code, starts from crossing maximum number of zeros sounds correct but it did not handle when their numbers are equal. That is the reason why your example is incorrect.
